woeip/apps/data_collection/models.py

```diff
@@ -0,0 +1,118 @@
+
+    def __str__(self):
+        crud = []
+        if self.can_create:
+            crud.append('create')
+        if self.can_read:
+            crud.append('read')
+        if self.can_update:
+            crud.append('update')
+        if self.can_delete:
+            crud.append('delete')
+
+        crud = '/'.join(crud)
+
+        return f"{self.name} ({crud})"
+
+
+class Participant(models.Model):
+    name = models.CharField(max_length=256)
+    email = models.CharField(max_length=256)
+    organization = models.ForeignKey(Organization, on_delete=models.SET_NULL, blank=True, null=True)
+    role = models.ForeignKey(RoleEnum, on_delete=models.SET_NULL, blank=True, null=True)
+
+    def __str__(self):
+        return f"{self.name} <{self.email}> {self.organization} {self.role}"
+
+
+class RouteEnum(models.Model):
+    """Named routes"""
+    name = models.CharField(max_length=256)
+    latlon = LineStringField()
+
+    def __str__(self):
+        return self.name
+
+
+class Session(models.Model):
+    """A single air quality outing. Can link to several SessionData, e.g., raw data files."""
+    date_collected = models.DateTimeField(auto_now_add=True)
+    route = models.ForeignKey(RouteEnum, on_delete=models.SET_NULL, blank=True, null=True)
+    collected_by = models.ForeignKey(Participant, on_delete=models.CASCADE)
+    metadata = JSONField(blank=True, null=True)
+
+    def __str__(self):
+        return f"{self.date_collected} {self.collected_by}"
+
+
+class Device(models.Model):
+    name = models.CharField(max_length=256)
+    manufacturer = models.CharField(max_length=256)
+    serial_number = models.CharField(max_length=256)
+    model_number = models.CharField(max_length=256)
+    calibration_date = models.DateField()
+    firmware_version = models.CharField(max_length=256)
+
+    def __str__(self):
+        return f"{self.name} {self.model_number} {self.serial_number}"
+
+
+class Sensor(models.Model):
+    """A sensor is something that measures something, i.e., it produces a single measurement value
+    at a time.
+    """
+    name = models.CharField(max_length=256)
+    unit = models.CharField(max_length=256,
+                            help_text="""Measurement unit, e.g., mg/m3, ppm, etc.""")
+    device = models.ForeignKey(Device, on_delete=models.CASCADE)
+
+    def __str__(self):
+        return f"{self.name} ({self.device.name})"
+
+
+class SessionData(models.Model):
+    """The raw data file generated during a session. Assumes one and only one file per sensor,
+    although multiple sensors can be linked to one session"""
+    uri = models.CharField(max_length=256)
+    sensor = models.ForeignKey(Sensor, on_delete=models.SET_NULL, blank=True, null=True)
+    session = models.ForeignKey(Session, on_delete=models.CASCADE)
+    upload_time = models.DateTimeField(auto_now_add=True)
+    uploaded_by = models.ForeignKey(Participant, on_delete=models.SET_NULL, blank=True, null=True)
+    hash = models.CharField(max_length=256)
```

Comment:
I'm assuming this is supposed to represent a computed hash. What are you hashing to generate this?

Reply:
I was thinking of this as a way to ensure that the same raw data file isn't uploaded twice. So hashing the content of the raw data file itself.

Reply:
Got it. I have a bunch of thoughts on this:

1. MD5 hashes are only 32 characters, so you can decrease `max_length`.
2. There is some [prior art](https://stackoverflow.com/questions/38256201/django-prevent-file-duplicates-in-view) you can use to override the `save` method to do this.
3. I recommend doing this work in a follow-on PR.